node/src/components/chain_synchronizer/operations.rs

```diff
@@ -351,125 +357,184 @@ impl CanUseSyncingNodes for Deploy {}
+        if new_peer_list.is_empty() && total_attempts % 100 == 0 {
             warn!(
-                attempts,
+                total_attempts,
+                attempts_after_bootstrapped,
+                has_connected_to_network,
                 item_type = ?T::TAG,
                 ?id,
                 can_use_syncing_nodes = %T::can_use_syncing_nodes(),
                 "failed to attempt to fetch item due to no fully-connected peers"
             );
         }
 
-        for peer in new_peer_list {
-            trace!(
-                "attempting to fetch {:?} with id {:?} from {:?}",
-                T::TAG,
+        if let Some(value) = fetch_from_peers(new_peer_list, id, ctx).await {
+            return value;
+        }
+
+        total_attempts += 1;
+        if has_connected_to_network {
+            attempts_after_bootstrapped += 1;
+        }
+        if attempts_after_bootstrapped >= ctx.config.max_sync_fetch_attempts() {
+            return Err(FetchWithRetryError::AttemptsExhausted {
                 id,
-                peer
-            );
-            match ctx.effect_builder.fetch::<T>(id, peer).await {
-                Ok(fetched_data @ FetchedData::FromStorage { .. }) => {
-                    trace!(
-                        "did not get {:?} with id {:?} from {:?}, got from storage instead",
-                        T::TAG,
-                        id,
-                        peer
-                    );
-                    return Ok(fetched_data);
-                }
-                Ok(fetched_data @ FetchedData::FromPeer { .. }) => {
-                    trace!("fetched {:?} with id {:?} from {:?}", T::TAG, id, peer);
-                    return Ok(fetched_data);
-                }
-                Err(FetcherError::Absent { .. }) => {
-                    warn!(
-                        ?id,
-                        tag = ?T::TAG,
-                        ?peer,
-                        "chain sync could not fetch; trying next peer",
-                    );
-                    ctx.mark_bad_peer(peer);
-                }
-                Err(FetcherError::TimedOut { .. }) => {
-                    warn!(
-                        ?id,
-                        tag = ?T::TAG,
-                        ?peer,
-                        "peer timed out",
-                    );
-                    ctx.mark_bad_peer(peer);
-                }
-                Err(error @ FetcherError::CouldNotConstructGetRequest { .. }) => return Err(error),
-            }
+                total_attempts,
+                attempts_after_bootstrapped,
+            });
         }
+
         tokio::time::sleep(ctx.config.retry_interval()).await;
-        attempts += 1;
     }
 }
 
+async fn fetch_from_peers<REv, T>(
+    new_peer_list: Vec<NodeId>,
+    id: <T as Item>::Id,
+    ctx: &ChainSyncContext<'_, REv>,
+) -> Option<Result<FetchedData<T>, FetchWithRetryError<T>>>
+where
+    T: Item + CanUseSyncingNodes + 'static,
```

Comment:
Do we still need `CanUseSyncingNodes` boundary?

Reply:
Nope, good catch. Removed in 46e6094a010468d7ca3ee16ef76aaf4c992040c2